File: game-server/src/main/java/work/fking/pangya/game/packet/outbound/HelloPacket.java
package work.fking.pangya.game.packet.outbound;

import io.netty.buffer.ByteBuf;
import io.netty.util.AttributeMap;
import work.fking.pangya.networking.crypt.PangCrypt;
import work.fking.pangya.networking.protocol.OutboundPacket;

public class HelloPacket implements OutboundPacket {

    private final byte[] unknown = {0x06, 0x00, 0x00, 0x3F, 0x00, 0x01, 0x01};
    private final int cryptKey;

    private HelloPacket(int cryptKey) {
        this.cryptKey = cryptKey;
    }

    public static HelloPacket create(int cryptKey) {

        if (cryptKey < 0 || cryptKey > PangCrypt.CRYPT_KEY_MAX) {
            throw new IllegalArgumentException("Invalid crypt key");
        }
        return new HelloPacket(cryptKey);
    }

    @Override
    public void encode(ByteBuf buffer, AttributeMap attributeMap) {
        buffer.writeByte(0);
        buffer.writeBytes(unknown);
        buffer.writeByte(cryptKey);
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/packet/outbound/HelloPacket.kt
package work.fking.pangya.game.packet.outbound

import io.netty.buffer.ByteBuf
import work.fking.pangya.networking.crypt.PangCrypt
import work.fking.pangya.networking.protocol.OutboundPacket

class HelloPacket(
    private val cryptKey: Int
) : OutboundPacket {

    init {
        require(!(cryptKey < 0 || cryptKey > PangCrypt.CRYPT_KEY_MAX)) { "Invalid crypt key" }
    }

    private val unknown = byteArrayOf(0x06, 0x00, 0x00, 0x3F, 0x00, 0x01, 0x01)

    override fun encode(target: ByteBuf) {
        target.writeByte(0)
        target.writeBytes(unknown)
        target.writeByte(cryptKey)
    }
}
